heal/vlmd/extract/utils.py

```diff
@@ -0,0 +1,293 @@
+        schema["properties"]["fields"]["items"]["properties"]
+    ).keys()
+    props_to_rearrange = (
+        set(root_names).intersection(field_names).difference(annotation_names)
+    )
+
+    propnames_exp = "|".join("^" + name + "$" for name in props_to_rearrange)
+    names_to_rearrange = [name for name in propnames if re.match(propnames_exp, name)]
+
+    return list(names_to_rearrange)
+
+
+def embed_data_dictionary_props(flat_fields, flat_root, schema):
+    """
+    Embed (flattened) root level props in each (flattened) field
+    if field level prop is missing but the field level prop exists
+    and is not an annotation property (title,description).
+
+    Args
+
+        flat_fields: array of dicts or pd.DataFrame (or something that can be converted into DataFrame)
+        flat_root: dict or pd.Series or something that can be converted into Series
+        schema: schema for determining what fields should be embedded (note, this is flattened in fxn)
+
+    Returns
+
+        pd.DataFrame with the flat fields with the embedded root properties
+    """
+    flat_fields = pd.DataFrame(flat_fields)
+    propnames = _get_propnames_to_rearrange(list(flat_root.keys()), schema)
+    flat_root = pd.Series(flat_root).loc[propnames]  # take out annotation props
+    if len(flat_root) > 0:
+        for propname in propnames:
+            if propname in flat_root:
+                if not propname in flat_fields:
+                    flat_fields.insert(0, propname, flat_root[propname])
+                else:
+                    flat_fields[propname].fillna(flat_root[propname], inplace=True)
+
+    return flat_fields
+
+
+def refactor_field_props(flat_fields, schema):
+    """
+    Given a flattened array of dicts corresponding to the unflattened schema,
+    move up (ie `refactor`) flattened properties that are both in the root
+    (ie table level; level up from field records) and in the fields.
+
+    """
+    flat_fields_df = pd.DataFrame(flat_fields)
+    propnames = set(
+        _get_propnames_to_rearrange(flat_fields_df.columns.tolist(), schema)
+    )
+    flat_record = pd.Series(dtype="object")
+    for name in propnames:
+        in_df = name in flat_fields_df
+        if in_df:
+            # need to handle if some values are pandas series
+            if isinstance(flat_fields_df[name], pd.DataFrame):
+                is_one_unique = (flat_fields_df[name].nunique() == 1).all()
+            elif isinstance(flat_fields_df[name], pd.Series):
+                if is_object_dtype(flat_fields_df[name]):
+                    is_one_unique = len(flat_fields_df[name].map(str).unique()) == 1
+                else:
+                    is_one_unique = flat_fields_df[name].nunique() == 1
+            if is_one_unique:
+                flat_record[name] = flat_fields_df.pop(name).iloc[0]
+                if isinstance(flat_record[name], pd.Series):
+                    flat_record[name] = flat_record[name].to_list()
```

Comment:
I think I wrote this block of logic in the old `healdata-utils` code to patch some urgent bugs. But I didn't get to the root of the problem, so it is more like a band-aid fix
What I really hope we can do is to figure out why `flat_record[name]` can have so many possible types (pd.DataFrame, pd.Series, etc.). I feel that we might not need to support dealing with that many different types of `flat_record[name]` here, if we can fix the logic upstream to make sure the value of `flat_record[name]` will only ends up in one type

Reply:
We can also create another ticket for what I said above, and put a `FIXME` note here

Reply:
With some testing I got a "custom" field into this block. It was always a pd.Series. I think any "custom" or "custom.<KEY>" can only be a single column in the CSV input file - so it is always a pd.Series. I could not get any variables to read in as a pd.DataFrame.